File: 2_add_two_numbers/add2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode(0, None)
        p, q, current = l1, l2, dummyHead
        carry = 0
        while p or q:
            x = p.val if p else 0
            y = q.val if q else 0
            sum = carry + x + y
            carry = int(sum/10)
            current.next = ListNode(sum % 10)
            current = current.next
            if p :
                p = p.next
            if q :
                q = q.next
        if carry>0 :
            current.next = ListNode(carry)
        return dummyHead.next

File: 2_add_two_numbers/add2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        ListNode* dummyHead = new ListNode();
        ListNode* current = dummyHead;
        int carry = 0;
        int x, y, sum;
        while ( l1 != NULL || l2 != NULL) {
            x = (l1 == NULL)? 0 : l1->val;
            y = (l2 == NULL)? 0 : l2->val;
            sum = carry + x + y;
            carry = sum/10;
            current->next = new ListNode(sum % 10);
            current = current->next;
            if (!l1) l1 = l1->next;
            if (!l2) l2 = l2->next;
        }
        if (carry>0) current->next = new ListNode(carry);
        return dummyHead->next;
    }
};